lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1264 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+#include "souper/Extractor/KLEEBuilder.h"
+#include "souper/Infer/InstSynthesis.h"
+
+using namespace souper;
+using namespace llvm;
+
+namespace {
+
+static cl::opt<bool> DebugSynthesis("souper-debug-synthesis",
+    cl::desc("Debug instruction synthesis (default=false)"),
+    cl::init(false));
+static cl::opt<int> CmdMaxCompNum("souper-synthesis-comp-num",
+    cl::desc("Maximum number of components (default=all)"),
+    cl::init(-1));
+static cl::opt<std::string> CmdUserCompKinds("souper-synthesis-comps",
+    cl::Hidden,
+    cl::desc("Comma-separated list of instruction synthesis component kinds"),
+    cl::init(""));
+static cl::opt<bool> IgnoreCost("souper-synthesis-ignore-cost",
+    cl::Hidden,
+    cl::desc("Ignore cand inst cost (default=false)"),
+    cl::init(false));
+
+}
+
+namespace souper {
+
+static const std::string INPUT_PREFIX = "in_";
+static const std::string LOC_PREFIX = "loc_";
+static const std::string LOC_SEP = "_";
+static const std::string COMP_INPUT_PREFIX = "compin_";
+static const std::string CONST_PREFIX = "const_";
+
+std::error_code InstSynthesis::synthesize(SMTLIBSolver *SMTSolver,
+                                          const BlockPCs &BPCs,
+                                          const std::vector<InstMapping> &PCs,
+                                          Inst *LHS, Inst *&RHS, InstContext &IC,
+                                          unsigned Timeout) {
+  std::error_code EC;
+  RHS = 0;
+
+  // The order is important
+  setCompLibrary();
+  initInputVars(LHS, IC);
+  setDefaultWidth(LHS);
+  addZSTComps(LHS);
+  initOutput(LHS, IC);
+  initComponents(IC);
+  initConstComponents(IC);
+  initLocations();
+
+  N = I.size();
+  M = Comps.size() + N;
+
+  int LHSCost = cost(LHS);
+
+  if (DebugSynthesis)
+    printInitInfo();
+
+  setInvalidWirings();
+
+  // Init a new set of path conditions
+  std::vector<InstMapping> WiringPCs;
+  addConstraints(WiringPCs, IC);
+
+  // Create the main wiring query (aka connectivity contraint)
+  Inst *WiringQuery = getConnectivityConstraint(IC);
+
+  // Initial concrete input set S.
+  // With every new input set that proves a synthesised program is invalid,
+  // we'll have to copy WiringQuery and replace its inputs with the new
+  // concrete values from S
+  std::vector<std::map<Inst *, Inst *>> S;
+
+  // Ask the solver for a couple of arbitrary initial inputs
+  auto InputPCs = PCs;
+  for (unsigned I = 0; I < 4; ++I) {
```

Comment:
Where does 4 come from?
Be careful about variables named I since we often use that to mean an Inst *.